File: Apps/ContactLogoAndroid/app/src/main/java/com/contactlogo/engine/CompanyCatalog.kt
package com.contactlogo.engine

/**
 * R8.3 — offline name -> official domain catalog, plus the CATALOG_TAIL_OK
 * sub-brand reduction ("H-E-B Pharmacy" -> heb.com, "Kroger Marketplace Cypress"
 * -> kroger.com) without folding trade-word tails onto the wrong parent
 * ("Delta Dental" MUST NOT reduce to delta.com).
 */
object CompanyCatalog {

    private val domains: Map<String, String> = mapOf(
        "apple" to "apple.com", "apple inc" to "apple.com",
        "google" to "google.com", "alphabet" to "google.com",
        "microsoft" to "microsoft.com", "amazon" to "amazon.com",
        "meta" to "meta.com", "facebook" to "facebook.com", "instagram" to "instagram.com",
        "netflix" to "netflix.com", "spotify" to "spotify.com", "uber" to "uber.com",
        "lyft" to "lyft.com", "airbnb" to "airbnb.com", "door dash" to "doordash.com",
        "doordash" to "doordash.com", "instacart" to "instacart.com",
        "walmart" to "walmart.com", "target" to "target.com", "costco" to "costco.com",
        "home depot" to "homedepot.com", "the home depot" to "homedepot.com",
        "lowes" to "lowes.com", "lowe's" to "lowes.com",
        "starbucks" to "starbucks.com", "mcdonalds" to "mcdonalds.com",
        "mcdonald's" to "mcdonalds.com", "chick fil a" to "chick-fil-a.com",
        "chick-fil-a" to "chick-fil-a.com", "chipotle" to "chipotle.com",
        "delta" to "delta.com", "delta air lines" to "delta.com",
        "united" to "united.com", "united airlines" to "united.com",
        "american airlines" to "aa.com", "southwest" to "southwest.com",
        "southwest airlines" to "southwest.com",
        "chase" to "chase.com", "jpmorgan" to "jpmorganchase.com",
        "bank of america" to "bankofamerica.com", "bofa" to "bankofamerica.com",
        "wells fargo" to "wellsfargo.com", "citi" to "citi.com", "citibank" to "citi.com",
        "american express" to "americanexpress.com", "amex" to "americanexpress.com",
        "capital one" to "capitalone.com", "discover" to "discover.com",
        "charles schwab" to "schwab.com", "schwab" to "schwab.com",
        "fidelity" to "fidelity.com", "vanguard" to "vanguard.com",
        "progressive" to "progressive.com", "geico" to "geico.com",
        "state farm" to "statefarm.com", "allstate" to "allstate.com",
        "at&t" to "att.com", "att" to "att.com", "verizon" to "verizon.com",
        "t-mobile" to "t-mobile.com", "tmobile" to "t-mobile.com",
        "comcast" to "xfinity.com", "xfinity" to "xfinity.com", "spectrum" to "spectrum.com",
        "fedex" to "fedex.com", "ups" to "ups.com", "usps" to "usps.com",
        "heb" to "heb.com", "h-e-b" to "heb.com", "kroger" to "kroger.com",
        "publix" to "publix.com", "walgreens" to "walgreens.com", "cvs" to "cvs.com",
        "trader joes" to "traderjoes.com", "trader joe's" to "traderjoes.com",
        "aldi" to "aldi.us", "whole foods" to "wholefoodsmarket.com",
        "kaiser" to "kp.org", "kaiser permanente" to "kp.org",
        "quest" to "questdiagnostics.com", "quest diagnostics" to "questdiagnostics.com",
        "labcorp" to "labcorp.com", "enterprise" to "enterprise.com",
        "hertz" to "hertz.com", "avis" to "avis.com",
        "shell" to "shell.com", "chevron" to "chevron.com",
        "exxon" to "exxon.com", "exxonmobil" to "exxonmobil.com",
        "bp" to "bp.com", "7-eleven" to "7-eleven.com",
        "wawa" to "wawa.com", "buc-ees" to "buc-ees.com",
        "h&r block" to "hrblock.com",
        "cisco" to "cisco.com",
        // MATCHING-ENGINE §2 rule 4 alias map.
        "txt" to "texasbytexas.com", "gcx" to "raise.com"
    )

    /** R8.3 CATALOG_TAIL_OK = GEO_WORDS (R6.6) union SUBBRAND_TAIL, tested word by word. */
    private val subbrandTailWords: Set<String> = setOf(
        "pharmacy", "deli", "bakery", "fuel", "gas", "market", "marketplace", "optical", "photo",
        "curbside", "drive", "thru", "corporate", "hq", "distribution", "warehouse"
    )
    private val tailOkWords: Set<String> = Normalize.GEO_WORDS_LITERAL + subbrandTailWords
    private val tailOkNumeric = listOf(Regex("""^#\d*$"""), Regex("""^\d{2,5}$"""))

    private fun isTailOkWord(w: String): Boolean {
        if (w.lowercase() in tailOkWords) return true
        return tailOkNumeric.any { it.matches(w) }
    }

    /**
     * R8.3 — may this tail be dropped, leaving the head brand?
     *
     * Two conditions, and both are needed.  Requiring only that *some* word be
     * tail-ok reduced "Delta Dental Center" to delta.com on the strength of
     * `center` alone — an airline's logo on a dental practice.  Requiring that
     * *every* word be tail-ok instead rejected "Walgreens Mason Rd", because a
     * street name is not on any list and never can be.
     *
     * So: something must positively mark the tail as a place or department, and
     * nothing in it may name a different trade.  An unrecognised word ("mason")
     * is tolerated as part of an address; an ORG_SIGNAL word ("dental") is not,
     * since it makes the tail a business of its own.  SUBBRAND_TAIL wins where
     * the two lists overlap — "pharmacy" and "bakery" are H-E-B departments.
     */
    fun isCatalogTailOK(tail: List<String>): Boolean {
        if (tail.isEmpty()) return false
        if (tail.none { isTailOkWord(it) }) return false
        return tail.all { isTailOkWord(it) || !Normalize.isOrgSignalWord(it) }
    }

    /**
     * R8.3: `DOMAINS[companyKey(raw)]`, else the space-collapsed key, else the
     * longest-head / sub-brand-tail reduction, else null.
     */
    fun domainForName(name: String): String? {
        val k = Normalize.companyKey(name)
        if (k.isEmpty()) return null
        domains[k]?.let { return it }

        val nospace = k.replace(" ", "")
        domains[nospace]?.let { return it }

        val words = k.split(" ")
        for (i in words.size - 1 downTo 1) {
            val head = words.subList(0, i).joinToString(" ")
            val tailWords = words.subList(i, words.size)
            val dom = domains[head] ?: domains[head.replace(" ", "")]
            if (dom != null && isCatalogTailOK(tailWords)) return dom
        }
        return null
    }
}

File: Sources/ContactLogoKit/Normalize/CompanyCatalog.swift
import Foundation

/// Offline name → domain table ported from `vendor/crest/src/lib/contacts.ts`.
/// Used when a contact has no usable website or work email. Homonyms still
/// go through the review-first confidence cap (MATCHING-ENGINE §4).
public enum CompanyCatalog {

    /// Lowercased, punctuation-stripped keys → registrable domain.
    static let domains: [String: String] = [
        "apple": "apple.com", "apple inc": "apple.com",
        "google": "google.com", "alphabet": "abc.xyz",
        "microsoft": "microsoft.com", "amazon": "amazon.com",
        "meta": "meta.com", "facebook": "facebook.com", "instagram": "instagram.com",
        "tesla": "tesla.com", "nvidia": "nvidia.com",
        "netflix": "netflix.com", "spotify": "spotify.com",
        "adobe": "adobe.com", "salesforce": "salesforce.com",
        "oracle": "oracle.com", "ibm": "ibm.com", "intel": "intel.com",
        "cisco": "cisco.com", "stripe": "stripe.com", "paypal": "paypal.com",
        "visa": "visa.com", "mastercard": "mastercard.com",
        "american express": "americanexpress.com", "amex": "americanexpress.com",
        "chase": "chase.com", "jpmorgan": "jpmorganchase.com",
        "jp morgan": "jpmorganchase.com", "jpmorgan chase": "jpmorganchase.com",
        "bank of america": "bankofamerica.com",
        "wells fargo": "wellsfargo.com",
        "citi": "citi.com", "citibank": "citi.com", "citigroup": "citi.com",
        "geico": "geico.com",
        "state farm": "statefarm.com", "state farm insurance": "statefarm.com",
        "allstate": "allstate.com", "allstate insurance": "allstate.com",
        "usaa": "usaa.com", "usaa insurance": "usaa.com",
        "verizon": "verizon.com",
        "at&t": "att.com", "att": "att.com",
        "t-mobile": "t-mobile.com", "tmobile": "t-mobile.com",
        "united airlines": "united.com", "united": "united.com",
        "american airlines": "aa.com",
        "delta": "delta.com", "delta air lines": "delta.com", "delta airlines": "delta.com",
        "southwest": "southwest.com", "southwest airlines": "southwest.com",
        "jetblue": "jetblue.com", "alaska airlines": "alaskaair.com",
        "fedex": "fedex.com", "ups": "ups.com", "usps": "usps.com",
        "the home depot": "homedepot.com", "home depot": "homedepot.com",
        "lowes": "lowes.com", "lowe's": "lowes.com",
        "costco": "costco.com", "walmart": "walmart.com", "target": "target.com",
        "starbucks": "starbucks.com",
        "mcdonalds": "mcdonalds.com", "mcdonald's": "mcdonalds.com",
        "uber": "uber.com", "lyft": "lyft.com", "doordash": "doordash.com",
        "airbnb": "airbnb.com", "nike": "nike.com",
        "samsung": "samsung.com", "sony": "sony.com",
        "ford": "ford.com", "bmw": "bmw.com",
        "centerpoint energy": "centerpointenergy.com", "centerpoint": "centerpointenergy.com",
        "x ai": "x.ai", "xai": "x.ai",
        "square": "squareup.com",
        "capital one": "capitalone.com", "discover": "discover.com",
        "intuit": "intuit.com", "turbotax": "turbotax.intuit.com",
        "quickbooks": "quickbooks.intuit.com",
        "h&r": "hrblock.com", "h&r block": "hrblock.com", "h and r block": "hrblock.com",
        "edward jones": "edwardjones.com",
        "charles schwab": "schwab.com", "td ameritrade": "tdameritrade.com",
        "etrade": "etrade.com", "robinhood": "robinhood.com", "coinbase": "coinbase.com",
        "american tower": "americantower.com", "crown castle": "crowncastle.com",
        "waste management": "wm.com",
        "republic": "republicservices.com", "republic services": "republicservices.com",
        "waste connections": "wasteconnections.com",
        "texas instruments": "ti.com", "qualcomm": "qualcomm.com",
        "broadcom": "broadcom.com", "amd": "amd.com",
        "advanced micro devices": "amd.com",
        "palantir": "palantir.com", "snowflake": "snowflake.com",
        "databricks": "databricks.com", "servicenow": "servicenow.com",
        "workday": "workday.com", "autodesk": "autodesk.com",
        "electronic arts": "ea.com", "activision": "activision.com",
        "take-two": "take2games.com", "roblox": "roblox.com", "unity": "unity.com",
        "epic games": "epicgames.com", "valve": "valvesoftware.com",
        "steam": "steampowered.com",
        "comcast": "xfinity.com", "xfinity": "xfinity.com", "spectrum": "spectrum.com",
        "progressive": "progressive.com", "liberty mutual": "libertymutual.com",
        "farmers": "farmers.com", "nationwide": "nationwide.com",
        "heb": "heb.com", "h-e-b": "heb.com",
        "kroger": "kroger.com", "randalls": "randalls.com", "safeway": "safeway.com",
        "publix": "publix.com", "walgreens": "walgreens.com", "cvs": "cvs.com",
        "best buy": "bestbuy.com", "macy's": "macys.com", "macys": "macys.com",
        "hertz": "hertz.com", "enterprise": "enterprise.com", "avis": "avis.com",
        "hilton": "hilton.com", "marriott": "marriott.com", "hyatt": "hyatt.com",
        "fidelity": "fidelity.com", "vanguard": "vanguard.com", "schwab": "schwab.com",
        "trader joes": "traderjoes.com", "trader joe's": "traderjoes.com",
        "aldi": "aldi.us", "whole foods": "wholefoodsmarket.com", "whole foods market": "wholefoodsmarket.com",
        "kaiser": "kp.org", "kaiser permanente": "kp.org", "quest": "questdiagnostics.com", "quest diagnostics": "questdiagnostics.com",
        "labcorp": "labcorp.com", "enterprise rent-a-car": "enterprise.com",
        "shell": "shell.com", "chevron": "chevron.com",
        "exxon": "exxon.com", "exxonmobil": "exxonmobil.com", "bp": "bp.com", "7 eleven": "7-eleven.com", "7-eleven": "7-eleven.com",
        "wawa": "wawa.com", "bucees": "buc-ees.com", "buc ees": "buc-ees.com", "buc-ees": "buc-ees.com",
        "reliant": "reliant.com", "pg&e": "pge.com",
        "duke energy": "duke-energy.com", "amtrak": "amtrak.com",
        "txt": "texasbytexas.com", "texas by texas": "texasbytexas.com",
        "gcx": "raise.com", "raise": "raise.com"
    ]

    /// R8.3 `CATALOG_TAIL_OK` — a location or sub-brand tail after a known
    /// brand ("Walgreens Mason Rd", "H-E-B Pharmacy").  Trade words such as
    /// "dental" are deliberately absent, so "Delta Dental" never reduces to
    /// delta.com.
    static func isAllowedTail(_ tail: String) -> Bool {
        WordLists.isCatalogTailOK(tail)
    }

    public static func domain(forName raw: String) -> String? {
        let key = NameNormalizer.companyKey(raw)
        guard !key.isEmpty else { return nil }
        if let d = domains[key] { return d }
        let nospace = key.replacingOccurrences(of: " ", with: "")
        if let d = domains[nospace] { return d }

        let words = key.split(separator: " ").map(String.init).filter { !$0.isEmpty }
        guard words.count >= 2 else { return nil }
        for i in stride(from: words.count - 1, through: 1, by: -1) {
            let head = words[..<i].joined(separator: " ")
            let tail = words[i...].joined(separator: " ")
            let hit = domains[head] ?? domains[head.replacingOccurrences(of: " ", with: "")]
            if let hit, isAllowedTail(tail) {
                return hit
            }
        }
        return nil
    }
}
